File: Pastura/Pastura/Engine/ErrorReadability.swift
import Foundation

/// Returns the most human-readable description of an arbitrary `Error`.
///
/// Prefers ``LocalizedError/errorDescription`` when the error conforms,
/// falling back to `String(describing:)` otherwise. Used at the Engine wrap
/// points where foreign errors (typically `LLMError` or validator throws)
/// are bridged into ``SimulationError``-typed associated-value strings. Using
/// `"\(error)"` there loses the inner message because it stringifies the
/// whole enum case (`generationFailed(description: "...")`) instead of just
/// the meaningful text.
///
/// Foundation-only — Engine layer dependency rules (depends on LLM + Models)
/// are preserved.
nonisolated func readableDescription(_ error: Error) -> String {
  if let localized = (error as? LocalizedError)?.errorDescription {
    return localized
  }
  return String(describing: error)
}

File: shared/engine/src/commonMain/kotlin/com/pastura/engine/ErrorReadability.kt
package com.pastura.engine

/**
 * Returns the most human-readable description of an arbitrary [Throwable].
 *
 * The Swift original prefers `LocalizedError.errorDescription` and falls back
 * to `String(describing:)`. Kotlin has no `LocalizedError`, so the equivalent
 * mapping is **prefer [Throwable.message], else [Throwable.toString]** — the
 * `message` carries the meaningful text (mirroring how
 * [SimulationException.messageFor] renders readable text into `.message`),
 * while `toString()` prepends the class name and is only the last resort for a
 * message-less throwable.
 *
 * Landed as infra: there is **no Kotlin consumer yet**. The Engine
 * error-bridge call sites that wrap foreign errors into [SimulationError]-typed
 * strings — the analogue of the Swift wrap points — are later-Wave freight.
 *
 * Swift original: `Pastura/Pastura/Engine/ErrorReadability.swift`.
 * Ported for the ADR-023 §6 Stage-3 Engine migration (#501).
 */
internal fun readableDescription(error: Throwable): String = error.message ?: error.toString()
